File: src/leetCodeProblems/PrefixSum/PathSumIII437.java
package leetCodeProblems.PrefixSum;

/**
 * LeetCode - https://leetcode.com/problems/path-sum-iii/submissions/
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class PathSumIII437 {

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int val) {
            this.val = val;
        }
    }

    int ansCount = 0;
    int targetSum = 0;

    HashMap<Integer, Integer> pathSumMap;

    public void pathSumUtil(TreeNode node, int pathSum) {

        if (node == null) {
            return;
        }

        pathSum += node.val;

        int prefixSum = pathSum-targetSum;

        if (pathSumMap.containsKey(prefixSum)) {
            ansCount += pathSumMap.get(prefixSum);
        }

        if (pathSumMap.containsKey(pathSum)) {
            pathSumMap.put(pathSum, pathSumMap.get(pathSum) +1);
        }
        else {
            pathSumMap.put(pathSum, 1);
        }

        System.out.println("--Prefix Sum ->" + prefixSum);
        System.out.println(pathSumMap);
        System.out.println("Ans --->" + ansCount);

        pathSumUtil(node.left, pathSum);
        pathSumUtil(node.right, pathSum);

        pathSumMap.put(pathSum, pathSumMap.get(pathSum)-1);
    }

    public int pathSum(TreeNode root, int k) {

        targetSum = k;
        ansCount = 0;

        pathSumMap = new HashMap<>();
        pathSumMap.put(0, 1);

        pathSumUtil(root, 0);

        return ansCount;
    }

    public static void main(String[] args) {

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);
        root.right.left = new TreeNode(2);
        root.right.right = new TreeNode(1);

        PathSumIII437 ob = new PathSumIII437();

        int output = ob.pathSum(root, 2); // o/p =

        System.out.println(output);

    }
}

File: src/leetCodeProblems/PrefixSum/PathSumIII437.go
package main

/*
- LeetCode - https://leetcode.com/problems/path-sum-iii/
- Time - O(n)
- Space - O(n)
*/

func pathSumUtil(node *TreeNode, pathSum, targetSum int, out *int, m map[int]int) {
	if node == nil {
		return
	}

	pathSum += node.Val
	prefixSum := pathSum - targetSum

	if v, ok := m[prefixSum]; ok {
		*out += v
	}

	m[pathSum]++

	pathSumUtil(node.Left, pathSum, targetSum, out, m)
	pathSumUtil(node.Right, pathSum, targetSum, out, m)

	m[pathSum]--
}

func pathSum(root *TreeNode, tSum int) int {

	m := make(map[int]int)
	m[0] = 1

	out := 0
	pathSumUtil(root, 0, tSum, &out, m)

	return out
}

// func main() {
// 	root := new(TreeNode) //Returns pointer to TreeNode object
// 	root.Val = 10
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 5
// 	root.Left.Left = new(TreeNode)
// 	root.Left.Left.Val = 3
// 	root.Left.Right = new(TreeNode)
// 	root.Left.Right.Val = 2
// 	root.Left.Right.Right = new(TreeNode)
// 	root.Left.Right.Right.Val = 1
// 	root.Left.Left.Left = new(TreeNode)
// 	root.Left.Left.Left.Val = 3
// 	root.Left.Left.Right = new(TreeNode)
// 	root.Left.Left.Right.Val = -2

// 	root.Right = new(TreeNode)
// 	root.Right.Val = -3
// 	root.Right.Right = new(TreeNode)
// 	root.Right.Right.Val = 11

// 	log.Println(pathSum(root, 8))

// 	// root := new(TreeNode) //Returns pointer to TreeNode object
// 	// root.Val = 5
// 	// root.Left = new(TreeNode)
// 	// root.Left.Val = 4
// 	// root.Right = new(TreeNode)
// 	// root.Right.Val = 8

// 	// root.Left.Left = new(TreeNode)
// 	// root.Left.Left.Val = 11
// 	// root.Left.Left.Left = new(TreeNode)
// 	// root.Left.Left.Left.Val = 7
// 	// root.Left.Left.Right = new(TreeNode)
// 	// root.Left.Left.Right.Val = 2

// 	// root.Right.Left = new(TreeNode)
// 	// root.Right.Left.Val = 13

// 	// root.Right.Right = new(TreeNode)
// 	// root.Right.Right.Val = 4

// 	// root.Right.Right.Left = new(TreeNode)
// 	// root.Right.Right.Left.Val = 5
// 	// root.Right.Right.Right = new(TreeNode)
// 	// root.Right.Right.Right.Val = 1

// 	//log.Println(pathSum(root, 22))
// }
